components/board/genericlinux/hw_pwm.go

```diff
@@ -0,0 +1,172 @@
+	files, err := os.ReadDir(chipDir)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, file := range files {
+		if strings.Contains(file.Name(), "pwmchip") && file.IsDir() {
+			chipPath := fmt.Sprintf("/sys/class/pwm/%s", file.Name())
+			return &pwmDevice{chipPath: chipPath, line: line}, nil
+		}
+	}
+	return nil, errors.Errorf("Could not find any PWM device with name %s", chipName)
+}
+
+func writeValue(filepath string, value uint64) error {
+	// The permissions on the file (the third argument) aren't important: if the file needs to be
+	// created, something has gone horribly wrong!
+	return os.WriteFile(filepath, []byte(fmt.Sprintf("%d", value)), 0o600)
+}
+
+func (pwm *pwmDevice) writeChip(filename string, value uint64) error {
+	return writeValue(fmt.Sprintf("%s/%s", pwm.chipPath, filename), value)
+}
+
+func (pwm *pwmDevice) writeLine(filename string, value uint64) error {
+	return writeValue(fmt.Sprintf("%s/pwm%d/%s", pwm.chipPath, pwm.line, filename), value)
+}
+
+// Export tells the OS that this pin is in use, and enables configuration via sysfs.
+func (pwm *pwmDevice) export() error {
+	if pwm.isExported {
+		return nil // Already exported
+	}
+	if err := pwm.writeChip("export", uint64(pwm.line)); err != nil {
+		return err
+	}
+	pwm.isExported = true
+	return nil
+}
+
+// Unexport turns off any PWM signal the pin was providing, and tells the OS that this pin is no
+// longer in use (so it can be reused as an input pin, etc.).
+func (pwm *pwmDevice) unexport() error {
+	if !pwm.isExported {
+		return nil // Already unexported
+	}
+	if err := pwm.disable(); err != nil {
+		return err
+	}
+	if err := pwm.writeChip("unexport", uint64(pwm.line)); err != nil {
+		return err
+	}
+	pwm.isExported = false
+	return nil
+}
+
+// Enable tells an exported pin to output the PWM signal it has been configured with.
+func (pwm *pwmDevice) enable() error {
+	if pwm.isEnabled {
+		return nil // Already enabled
+	}
+	if err := pwm.writeLine("enable", 1); err != nil {
+		return err
+	}
+	pwm.isEnabled = true
+	return nil
+}
+
+// Disable tells an exported pin to stop outputting its PWM signal, but it is still available for
+// reconfiguring and re-enabling.
+func (pwm *pwmDevice) disable() error {
+	if !pwm.isEnabled {
+		return nil // Already disabled
+	}
+	if err := pwm.writeLine("enable", 0); err != nil {
+		return err
+	}
+	pwm.isEnabled = false
+	return nil
+}
+
+// SetPwm configures an exported pin and enables its output signal.
+// Warning: if this function returns a non-nil error, it could leave the pin in an indeterminate
+// state. Maybe it's exported, maybe not. Maybe it's enabled, maybe not. The new frequency and duty
+// cycle each might or might not be set.
+func (pwm *pwmDevice) SetPwm(freqHz uint, dutyCycle float64) error {
+	pwm.mu.Lock()
+	defer pwm.mu.Unlock()
+
+	// What we really want in this function is a monad: for every interaction with sysfs, check if
+	// it returned an error, and if so return early.
```

Comment:
This comment might be too pretentious. As a Haskell/Rust/SML fan, it makes perfect sense to me, but I wonder if most people will think this comment raises more questions than it answers. If so, I can remove it.